src/domain/tradingview/TVDataProvider.ts

```diff
@@ -0,0 +1,183 @@
+import { LAST_BAR_REFRESH_INTERVAL, SUPPORTED_RESOLUTIONS } from "config/tradingview";
+import { getLimitChartPricesFromStats, getStablePriceData, timezoneOffset } from "domain/prices";
+import { CHART_PERIODS, USD_DECIMALS } from "lib/legacy";
+import { formatAmount } from "lib/numbers";
+import { Bar } from "./types";
+import { formatTimeInBarToMs, getCurrentCandleTime } from "./utils";
+import { fillBarGaps, getCurrentPriceOfToken, getTokenChartPrice } from "./requests";
+import { BigNumberish } from "ethers";
+import { PeriodParams } from "charting_library";
+
+const initialHistoryBarsInfo = {
+  period: "",
+  data: [],
+  ticker: "",
+};
+
+export class TVDataProvider {
+  lastBar: Bar | null;
+  startTime: number;
+  lastTicker: string;
+  lastPeriod: string;
+  barsInfo: {
+    period: string;
+    data: Bar[];
+    ticker: string;
+  };
+
+  constructor() {
+    this.lastBar = null;
+    this.startTime = 0;
+    this.lastTicker = "";
+    this.lastPeriod = "";
+    this.barsInfo = initialHistoryBarsInfo;
+  }
+
+  async getCurrentPriceOfToken(chainId: number, ticker: string): Promise<BigNumberish> {
+    return getCurrentPriceOfToken(chainId, ticker);
+  }
+
+  async getTokenLastBars(chainId: number, ticker: string, period: string, limit: number): Promise<Bar[]> {
+    return getLimitChartPricesFromStats(chainId, ticker, period, limit);
+  }
+
+  async getTokenHistoryBars(
+    chainId: number,
+    ticker: string,
+    period: string,
+    periodParams: PeriodParams,
+    shouldRefetchBars: { current: boolean }
+  ): Promise<Bar[]> {
+    const barsInfo = this.barsInfo;
+    if (
+      !barsInfo.data.length ||
+      barsInfo.ticker !== ticker ||
+      barsInfo.period !== period ||
+      shouldRefetchBars.current
+    ) {
+      try {
+        const bars = await getTokenChartPrice(chainId, ticker, period);
```

Comment:
i think i will need to override getTokenChartPrice for synthetics here, maybe make it as a class method to?